wallet-new/src/Cardano/Wallet/Kernel/PrefilterTx.hs

```diff
@@ -137,22 +138,29 @@ prefilterTx wKey tx = ((prefInps,prefOuts'),metas)
 
         prefOuts' = Map.map (extendWithSummary (onlyOurInps,onlyOurOuts))
                             prefOuts
+        -- this Set.map does not change the number of elements because TxIn's are unique.
+        prefInps' = map (Set.map fst) prefInps
 
-        allAccounts = toList $ Map.keysSet prefInps <> Map.keysSet prefOuts
+        (prefInCoins  :: (Map HdAccountId Coin)) = map (sumCoinsUnsafe . map snd . Set.toList) prefInps
+        (prefOutCoins :: (Map HdAccountId Coin)) = map (\mp -> sumCoinsUnsafe $ map (toCoin . fst) mp) prefOuts'
 
-        metas = map (toTxMeta tx) allAccounts
+        allAccounts = toList $ Map.keysSet prefInps' <> Map.keysSet prefOuts
+        metas = map (\acc -> resolvedToTxMeta tx
+            (nothingToZero acc prefInCoins)
+            (nothingToZero acc prefOutCoins)
+            (onlyOurInps && onlyOurOuts) acc) allAccounts
 
 -- | Prefilter inputs of a transaction
```

Comment:
I think this whole prefiltering code needs a tidying pass, comments added, naming changed, etc. but we can take a look at that another time.